[CORE-8933] - Port partition_translator into scheduler interface

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR heavily refactors `datalake::translation::partition_translator` to work with the new scheduler interface. The end result is more modular, with many of the cluster-facing interfaces split out into shim interfaces. This should make fine-grained unit & fixture testing of the translator much easier, though we're deferring additional tests to a subsequent PR in the interest of unblocking dependent work, particularly the backpressure controllers.

This PR also introduces `partition_translator::current_lag_ms()` which calculates an heuristic approximation of translation lag for a given translator, namely `timestamp_clock::now() - last_translated_timestamp`, where `last_translated_timestamp` is a rough lower bound for the arrival time of the highest translated offset. This information is stored alongside last_translated_offset in translation_stm.

Known limitations:
- We're not doing lag interval tracking yet. Needs design and scoping (e.g. should we persist active intervals as partition state in the coordinator?)
- We call finish on each iteration of the translation loop, which flushes the write buffers and uploads files to S3. In terms of batching, this results in the same behavior as the v1 translator. Ideally we should add some logic to `finish` less frequently, perhaps based on some threshold or when nearing the end of a time slice. Needs design.

TODO:
- [x] remove the old partition translator from the tree and rename the new one.

@bharathv's original PR: https://github.com/redpanda-data/redpanda/pull/25077

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
datalake/t: Port partition_probe updates from v1 translator

Also changes the semantics of the lag_metrics ducktape test. In the new
scheduler regime, scheduled translators may sit idle waiting for translatable
records to appear on the input topic. This waiting may cross a commit interval
boundary, in which case commit lag metrics will not be updated until _after_
new records appear, waking up the translator. Only once those records are
processed (and the translator reconciled with the state of the coordinator),
will the commit lag metrics update to reflect the most recent state. In that
sense, commit lag metrics may tend to slightly overstate the case, particularly
in situations where an input topic is quiet for extended periods.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake/t: Port flush hook notifications from translator v1

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
config: Introduce cluster configs for datalake scheduler and upload

- datalake_scheduler_block_size
- datalake_scheduler_max_concurrent_translations
- datalake_scheduler_time_slice_s

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake/t: integrate target_lag_ms topic property & config

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake: Wire up scheduler configs

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake/t: Introduce deps for v2 partition_translator

- class writer_reservations_impl
- class coordinator_api
- class data_source
- class translation_context
- class coordinator_table_creator

Along with default implementations and factory functions. Finer grained testing
to come later.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake/mux: Introduce API for buffered & flushed bytes

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat
release
skip-units

oleiman commented:
/ci-repeat 1
release
skip-units

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62322-->
<!--998d16aa-5d00-4343-8478-cbd403b490a2-->
<!--title-->
## Retry command for [Build#62322](https://buildkite.com/redpanda/redpanda/builds/62322)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c4f48ad4-809a-4519-bcd1-b23b5c514eb8-->
<!--TEST RESULTS-->
<!--3769ef44-35cd-4c72-aede-56173e5d5413-->
<!--TEST RESULTS-->
<!--c4334076-1310-48c0-a29f-c5f7baffa137-->
<!--TEST RESULTS-->
<!--343093cd-f517-43bc-a6cb-9e42bde2833c-->
<!--TEST RESULTS-->
<!--dcc2b39e-6b1d-4005-954f-1783b609dfed-->
<!--TEST RESULTS-->
<!--2f46063f-36fd-4a5f-afdc-8ba17c31a14b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62322">build#62322</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c2-4877-91c3-1e0554950acb | FLAKY | 1/2 |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c3-4177-8159-d058d9343d81 | FAIL | 0/20 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c3-4177-8159-d058d9343d81 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c0-41fb-abda-16f5589aef08 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c2-4877-91c3-1e0554950acb | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c0-41fb-abda-16f5589aef08 | FLAKY | 1/4 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62322#01954116-48c3-4177-8159-d058d9343d81 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62375">build#62375</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_remount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62375#01954547-575e-4183-beda-9ca626df3507 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62375#01954547-575e-4183-beda-9ca626df3507 | FAIL | 0/20 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62375#01954547-5760-4ab5-9461-4041dc52f0ff | FLAKY | 1/4 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62375#01954547-575e-4183-beda-9ca626df3507 | FLAKY | 1/3 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62375#01954547-5760-4ab5-9461-4041dc52f0ff | FLAKY | 1/4 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62375#01954547-5761-493b-acaa-ca7dcdbeb41d | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62389">build#62389</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62389#0195462c-b8f3-415f-8220-90d0f857c525 | FLAKY | 1/2 |
| rptest.tests.acls_test.AccessControlListTestUpgrade.test_describe_acls.use_tls=False.use_sasl=False.enable_authz=True.authn_method=mtls_identity.client_auth=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fc-443e-84d2-1110ed0e0c70 | FAIL | 0/1 |
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#0195468b-6098-4f82-8b43-d54db0dce538 | FLAKY | 1/2 |
| rptest.tests.datalake.3rdparty_maintenance_test.Datalake3rdPartyMaintenanceTest.test_e2e_basic.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fb-4cfa-bf98-c2c55d6b2117 | FAIL | 0/20 |
| rptest.tests.datalake.coordinator_retention_test.CoordinatorRetentionTest.test_retention.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fb-4cfa-bf98-c2c55d6b2117 | FAIL | 0/20 |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_dlq_table_for_invalid_records.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fb-4cfa-bf98-c2c55d6b2117 | FAIL | 0/20 |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_invalid_record_action.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_topic_property=True.action=dlq_table | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fb-4cfa-bf98-c2c55d6b2117 | FAIL | 0/20 |
| rptest.tests.datalake.recovery_mode_test.DatalakeRecoveryModeTest.test_recovery_mode.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fb-4cfa-bf98-c2c55d6b2117 | FAIL | 0/20 |
| rptest.tests.datalake.recovery_mode_test.DatalakeRecoveryModeTest.test_recovery_mode.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fc-443e-84d2-1110ed0e0c70 | FAIL | 0/20 |
| rptest.tests.datalake.simple_connect_test.RedpandaConnectIcebergTest.test_translating_avro_serialized_records.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fc-443e-84d2-1110ed0e0c70 | FAIL | 0/20 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#0195468b-6096-4581-8f8f-078611c28282 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=2.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#0195468b-6097-4371-977e-9077ad1f743f | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#0195468b-6096-4581-8f8f-078611c28282 | FLAKY | 1/4 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#0195468b-6098-4f82-8b43-d54db0dce538 | FLAKY | 1/10 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#0195468b-6096-4581-8f8f-078611c28282 | FLAKY | 1/8 |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all.use_transactions=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62389#01954688-f6fc-443e-84d2-1110ed0e0c70 | FAIL | 0/1 |
| translator_test_rpfixture.translator_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62389#0195462c-b8f3-415f-8220-90d0f857c525 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62474">build#62474</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_remount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62474#01954ec9-d4e0-40ab-9e27-7924ffc6d59a | FAIL | 0/20 |
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_unmount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62474#01954ec9-d4dd-4974-92fd-9777cf24a413 | FAIL | 0/20 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62474#01954ec9-d4dd-4974-92fd-9777cf24a413 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62474#01954ec9-d4e0-40ab-9e27-7924ffc6d59a | FLAKY | 1/4 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62474#01954ec9-d4de-4778-a172-4725a1c36e8f | FLAKY | 1/3 |
| translator_test_rpfixture.translator_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62474#01954e6a-dd34-4af0-bce0-ed32dd9c87d5 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62522">build#62522</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591b-9413-4ab9-a13e-33a426dfed62 | FLAKY | 1/2 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591b-9412-4d90-8e7e-7af91981b986 | FAIL | 0/20 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591b-9414-43c9-a1fd-dd28fbfde8e5 | FLAKY | 1/2 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591f-19a8-478b-8665-e27914f0aab2 | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591f-19a9-427a-a539-266f72041feb | FLAKY | 1/6 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591f-19a7-4ecb-867d-03221e1dd6c3 | FLAKY | 1/6 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591f-19a9-427a-a539-266f72041feb | FLAKY | 1/3 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591f-19a7-4ecb-867d-03221e1dd6c3 | FLAKY | 1/2 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets | ducktape | https://buildkite.com/redpanda/redpanda/builds/62522#0195591b-9413-4ab9-a13e-33a426dfed62 | FLAKY | 1/2 |
| translator_test_rpfixture.translator_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62522#019558d7-6f46-4cf2-8cbb-5dde7450cb13 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62577">build#62577</a></summary>


| test_id                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit        | https://buildkite.com/redpanda/redpanda/builds/62577#01955f5e-cd11-4362-9710-7af57a5114b5 | FLAKY         | 1/2      |
</details>
<!--/body-->

oleiman commented on src/v/datalake/translation/partition_translator_v2.cc:
```diff
@@ -0,0 +1,339 @@
+            continue;
+        }
+
+        // TODO: add logic to skip finish on every iteration. This is
+        // predicated on whether we are close to the lag deadline or if too many
+        // bytes have accumulated
+        // With out this, we finish() on every translation which makes it
+        // equivalent to the logic without this patch.
```
big todo

oleiman commented on src/v/datalake/translation/partition_translator_v2.cc:
```diff
@@ -0,0 +1,339 @@
+}
+
+scheduling::translation_status partition_translator::status() const {
+    // Biggest TODO: Currently we do not track intervals of target lag.
+    // So the scheduler is pretty much scheduling on defaults
+    // Track intervals return the correct remaining duration / memory
+    // reservations. The latter is already exposed as mux::buffered_bytes()
```
big todo

oleiman committed:
datalake: Add last_translated_timestamp to translation_stm

Intended to be a lower bound for the arrival time of the highest translated
offset. When we find untranslated records on a data source, we record the
current system time and the max translatable offset as a 'target'.
When translation meets or exceeds the cached target offset, replicate the
recorded timestamp. If the highest translated offset _exceeds_ the target, then
this will be an underestimate, which should allow us to _over_ estimate
instantaneous lag, hopefully improving our chances at achieving best-effort
translation deadlines.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
datalake/t: Introduce translator::current_lag_ms

Return the difference between system_clock::now and last_catchup_timestamp.

This is intended for use by translation throttling controllers, as a rough
approximation of the current translation lag for a given partition.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
⚠️  Note to reviewers - this PR is not quite ready to merge, I think, but given the time crunch probably best to get some eyes on its structure in case there's something obvious I'm missing ⚠️

oleiman commented:
/ci-repeat 1
release
skip-units

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62375-->
<!--da9e9e59-96bd-40ec-b56d-78e6a423b607-->
<!--title-->
## Retry command for [Build#62375](https://buildkite.com/redpanda/redpanda/builds/62375)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_chunked_compaction":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

oleiman commented:
CI Failure:
- [ducktape-release](https://buildkite.com/redpanda/redpanda/builds/62375) - `RandomNodeOperationsTest`, unrelated I think

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,441 @@
+}
+
+void writer_reservations_impl::update_current_memory_usage(size_t used_bytes) {
+    vassert(
```
What guarantees this? If this fires, where should we look?

oleiman replied:
this is only called from `serde_parquet_writer::add_data_struct`, after a row has been written, to return the number of flushed bytes to the memory tracker.

I'm reading it as an invariant enforced internally to the memory tracker, namely that it should never dish out more bytes than its total reserved memory. sound reasonable?

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,441 @@
+    _reservations.clear();
+}
+
+// Creates or alters the table by delegating to the coordinator.
+class coordinator_table_creator final : public table_creator {
```
This looks like a copy of code in datalake/translation/partition_translator.cc. Can we pull it out into its own header/cc if it's generally useful?

oleiman replied:
We're not going to keep both implementations. Removing the old version is on my list of todos.

bharathv replied:
ya i didn't remove the old impl because the code wasn't fully passing tests, we should probably do  to avoid confusion.

oleiman replied:
done locally, no worries. currently working on some code comments but will push again this evening.

oleiman replied:
done

andrwng commented on src/v/datalake/translation/partition_translator_v2.h:
```diff
@@ -0,0 +1,78 @@
+    ss::gate _gate;
+    ss::abort_source _as;
+
+    std::optional<inflight_translation_state> _inflight_translation_state;
+    ss::condition_variable _ready_to_translate;
```
nit: these could use some comments. Also maybe translate_until_stopped() and translate_when_notified(), particularly with any expectations we have of the scheduler

oleiman replied:
comments are WIP

oleiman replied:
done

andrwng commented on src/v/datalake/translation/deps.h:
```diff
@@ -21,4 +33,137 @@ class noop_mem_tracker : public writer_mem_tracker {
+
+    virtual model::term_id term() const = 0;
+
+    virtual ss::future<kafka::offset> wait_for_data_to_translate(
+      std::optional<kafka::offset> last_translated_offset, ss::abort_source&)
+      = 0;
```
nit: some comments here would be great -- it's not immediately clear what the returned offset is

oleiman replied:
done

andrwng commented on src/v/datalake/translation/partition_translator_v2.h:
```diff
@@ -0,0 +1,78 @@
+
+namespace datalake::translation {
+
+class partition_translator : public scheduling::translator {
```
nit: I know there are plenty of comments in scheduling.h, but I think it'd be helpful to add a comment with details about its expected lifecycle with respect to the scheduler and the underlying partition

oleiman replied:
done

andrwng commented on src/v/datalake/translation/deps.h:
```diff
@@ -21,4 +33,137 @@ class noop_mem_tracker : public writer_mem_tracker {
+      make_default_data_source(ss::lw_shared_ptr<cluster::partition>);
+};
+
+class translation_context {
+public:
```
nit: just a thought, maybe this should be called `partition_translator`, and the current `partition_translator` could be called `translation_runner` or somesuch. At least, I feel like that captures better that the guts of this class focus on the mechanics of scheduling, and the `scheduling::translator` impl's guts are more focused on running translation

oleiman replied:
i don't have a strong opinion about this beyond the principle of inertia. that said, I don't mind changing it around if you feel strongly. or we can do it later 😇

andrwng commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -399,4 +399,7 @@ def test_lag_metrics(self, cloud_storage_type):
-            self.wait_for_lag(m, DatalakeMetricsTest.commit_lag, 0)
+
+            time.sleep(5)
+            dl.produce_to_topic(self.topic_name, 1, msg_count=count)
+            self.wait_for_lag(m, DatalakeMetricsTest.commit_lag, count)
```
Could we end up committing so fast that we don't see the lag? Especially if we've reduced the commit interval?

oleiman replied:
> could we end up committing so fast that we don't see the lag?

No, but for a weird reason. I put a note about this somewhere (commit message maybe), but based on my reading this seems to be a fundamental limitation of the new scheduling loop.

Basically if we `wait_for_data_to_translate` on an idle topic before the last batch of files are committed (and `mark_file_committed_update` applied), we won't see the updated `last_committed_offset` until we re-enter the translation loop.

So the reason we always see the lag here is that the lag sticks to `count` basically indefinitely (assuming we don't produce any more records). It would be better if this weren't the case, but I don't see an obvious solution. Any ideas?

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -3916,6 +3916,41 @@ configuration::configuration()
+      {.needs_restart = needs_restart::yes, .visibility = visibility::tunable},
+      30s,
+      {.min = 1s, .max = 60s})
+  , iceberg_data_path_prefix(
+      *this,
+      "iceberg_data_path_prefix",
+      "Path prefix for uploading parquet files to remote object storage. "
+      " TODO: is it safe to set this after iceberg has been switched on?",
```
What's this for? Why not just hardcode `data` and `metadata`? Should we have one for metadata too?

>TODO: is it safe to set this after iceberg has been switched on

Is there a specific concern? It doesn't smell particularly unsafe

oleiman replied:
> What's this for?

Added it in response to a TODO in Bharath's PR that a block of static values should/could be configurable. I don't see an obvious reason for this one, so it might have been lumped in by mistake.

> specific concern?

Not really. I see a remote path, don't fully understand the implications of changing it on a live system, so I have to ask. Probably we can just remove this config though. @bharathv?

bharathv replied:
ya we can remove.. I didn't mean to make it a configurable (perhaps the code just got lumped under a TODO that caused this confusion, apologies)

oleiman replied:
done

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -3916,6 +3916,41 @@ configuration::configuration()
+  , datalake_scheduler_block_size_bytes(
+      *this,
+      "datalake_scheduler_block_size_bytes",
+      "Size, in bytes, of a memory block reserved for record translation, as "
+      "tracked by the datalake scheduler.",
```
nit: s/a/each? As is, it reads like all translation will reserve a single block of 4MiB

oleiman replied:
yeah fair point

oleiman replied:
done

andrwng commented on src/v/config/configuration.cc:
```diff
@@ -3916,6 +3916,41 @@ configuration::configuration()
+      "Time given, in milliseconds, to a datalake translation as scheduled by "
+      "the datalake scheduler. Once a translation is scheduled, it will run "
+      "until either a) the time slice has elapsed or b) all pending records "
+      "on its souirce partition have been translated.",
```
nit: source

oleiman replied:
done

andrwng commented on src/v/datalake/coordinator/coordinator.h:
```diff
@@ -77,6 +77,7 @@ class coordinator {
     struct last_offsets {
         std::optional<kafka::offset> last_added_offset;
         std::optional<kafka::offset> last_committed_offset;
+        std::optional<model::timestamp> last_added_timestamp;
```
Is this the Kafka data batch timestamp? Or something else? If Kafka, we might want to consider using a different timestamp, if this is going to be used for deadline tracking. Users can supply totally bogus timestamps and I can imagine that would mess with scheduling in unpredictable ways

I know we've chatted about other options, one that isn't great but might be good enough for our needs is something like time since translating to the HWM

oleiman replied:
i just shoved whatever was available already in the mux to expedite the mechanical changes. but yeah, right now it's

```c++
batch.header().first_timestamp.value() + record.timestamp_delta()
```

> Users can supply totally bogus timestamps and I can imagine that would mess with scheduling in unpredictable ways

This is intended as feedback for the @mmaslankaprv's throttling controller, but I assume same reasoning applies.

> time since translating to the HWM

sorry, what is "HWM"?

andrwng replied:
>sorry, what is "HWM"?

The high watermark, or I guess in this case I should've said last stable offset / max translatable offset

oleiman replied:
ah yeah, sorry i read it as a destination. you mean 'time point when we reached max translatable', yeah? that makes some sense. effectively any offset that becomes translatable after that is "lagging".

put another way, maybe we can store off  'wall clock time of last translation', which sort of decays to what your describing over subsequent updates, right? 

thought: none of these really accounts for idle topics. the quantity we care about is the time since the oldest untranslated offset became translatable, so any value we derive from the last translated offset gets noisy if the source topic goes quiet.

andrwng replied:
>thought: none of these really accounts for idle topics

Yeah, I guess in my mind, accounting for idle topics can/should be separated from deadline tracking. Truly idle topics won't even be considered by the scheduler (no work to do), but _kind of_ idle topics will be considered and should just be scheduled at a lower priority, no?

If the question is more about how throttling should account for it, I'm thinking could we use a similar idea, where topics that are fully idle (have translated to the max translatable) have 0 lag always? Or is there another consideration I'm missing?

oleiman replied:
ah, oops. yeah, had a couple of translator operations out of sequence in my head. this sounds right to me on both counts.

oleiman replied:
done. please have a look

andrwng commented on src/v/datalake/coordinator/coordinator.cc:
```diff
@@ -494,6 +494,7 @@ coordinator::sync_add_files(
     }
 
     auto added_last_offset = entries.back().last_offset;
+    auto added_last_timestamp = entries.back().last_timestamp;
```
>help with reconciliation.

Can you elaborate? And maybe add some comments in the corresponding state machines on how they are expected to be used

oleiman replied:
dropped this bit of state from the coordinator, so it lives in the translation_stm only

andrwng commented on src/v/datalake/translation/partition_translator_v2.cc:
```diff
@@ -323,6 +323,16 @@ scheduling::translation_status partition_translator::status() const {
+    if (!ts.has_value()) {
+        co_return std::chrono::milliseconds{0};
+    }
+    co_return std::chrono::milliseconds{
+      (model::timestamp::now() - ts.value()).value()};
```
Should we handle this being negative?

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -181,9 +181,22 @@ class partition_data_source : public data_source {
-        // todo: add logic to wait for enough data to translate.
-        // currently we just break even if a single batch of data is available
-        while (!has_more_data_to_translate(last_translated_offset)) {
+        // naively wait a few polling periods after some data appears, hopefully
```
I'm wondering how problematic this is in practice and how much it makes the difference, compared to e.g. improving the scheduling policy. Conceptually it feels like if there is even a single batch, there's no harm in telling the scheduler about it as long as the scheduler prioritizes appropriately w.r.t size, deadlines, etc

oleiman replied:
yeah that's a fair point. we're still going to wait to be scheduled in, so by the time we wake up there might be more data there anyway. @bharathv does the default scheduling policy take that into account?

bharathv replied:
ya I thought about this, my intent with the scheduler/policy API  was to keep it minimally stateful and give more control to translators as to when to translate (since they have most context about batching etc) which is what resulted in this API.  Otherwise there is one more dimension the scheduling policy has to evaluate making it a bit more complicated than it already is. That said one can also argue that things could be done the way as described in Andrew's comment

oleiman replied:
https://github.com/redpanda-data/redpanda/pull/25152#discussion_r1976681961

andrwng reviewed (commented):
Looks pretty good, it seems to make sense. The real test will be actually plugging it in I guess :)

mmaslankaprv commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,519 @@
 
+ss::future<> writer_reservations_impl::maybe_reserve_memory(
+  size_t bytes, ss::abort_source& as) {
+    while (_available_memory < bytes) {
```
is it possible that that `_available_memory` will always be smaller than `byte` ?

oleiman replied:
you mean in a situation where `bytes` exceeds the total memory in `reservations_tracker`? I think in that case 
- the scheduler gets notified of resource exhaustion
- the reservations_tracker::reserve_memory call is put into abortable sleep
- the scheduler stops all translators, each of which triggers its inflight abort source.
- which causes reserve_memory to throw sleep aborted
- which breaks the loop and lands in parquet_writer, which fails gracefully

so I don't _think_ so? do you see another outcome?

nvartolomei replied:
Can this lead to a case where nothing makes progress? I.e. N translations are scheduled, all of them try to acquire memory and do some writing but none of them ever acquire enough memory to complete a write? I see that on `maybe_reserve_memory` we just fail the write and restart.

bharathv replied:
In that case scheduler kicks in and aborts the translator using most memory so others can make progress?

nvartolomei replied:
Ah. @oleiman said it a bit differently

> the scheduler stops **all** translators, each of which triggers its inflight abort source.

The way you put it makes a bit more sense.

oleiman replied:
hmm, yeah good shout. not sure where I got that from. overall point being that `reservations_tracker::reserve_memory` will notify the scheduler if it can't service the reservation - maybe that should be captured in a comment somewhere?

mmaslankaprv commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,519 @@
+        if (_partition->is_read_replica_mode_enabled()) {
+            return model::offset_cast(_partition_proxy->start_offset());
+        }
+        return model::offset_cast(_partition_proxy->local_start_offset());
```
I think we should always use `_partition_proxy->start_offset()`. In case of node pool migration with fast partition movements we may introduce a gap as we may have to read from tiered storage

oleiman replied:
I'll have to take your word for it, unfortunately. Will include in the next push, but @bharathv please chime in if you feel the need.

andrwng replied:
Hm.. I'm curious how difficult it would be to address this in partition movement? Changing this here would have pretty major repercussions for how we handle topic recovery and WCR (we've opted to not backup coordinator state, and rely on local-only translation to avoid re-translating the entire log)

oleiman replied:
for posterity: discussed in group and concluded that it's not an issue in practice because we take this path iff we're starting up with no coordinator state. BUT it still might be a good idea to make fast partition movement contingent on max translatable offset.

mmaslankaprv commented on src/v/datalake/translation/deps.h:
```diff
@@ -21,4 +34,151 @@ class noop_mem_tracker : public writer_mem_tracker {
+    make_default_coordinator_api(coordinator::frontend&);
+};
+
+class data_source {
```
a comment on top of this class would be great

oleiman replied:
yup, comments are WIP

oleiman replied:
done

mmaslankaprv commented on src/v/datalake/translation/partition_translator_v2.h:
```diff
@@ -0,0 +1,79 @@
+#include "utils/retry_chain_node.h"
+
+namespace datalake::translation {
+
```
Why do we need `partition_translator_v2` ? Would it be possible to just change the v1 ?

oleiman replied:
yeah, see TODO in PR comment 🙂

oleiman replied:
done

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -59,8 +62,14 @@ ss::future<> translation_stm::do_apply(const model::record_batch& batch) {
           _log.trace,
-          "updating highest translated offset to {}",
-          value.highest_translated_offset);
+          "updating highest translated offset to {}[]",
```
missing placeholder for timestamp

oleiman replied:
removed this log

oleiman replied:
timestamp semantics changed, removed from log line

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -59,8 +62,14 @@ ss::future<> translation_stm::do_apply(const model::record_batch& batch) {
+          "updating highest translated offset to {}[]",
+          value.highest_translated_offset,
+          value.highest_translated_timestamp);
+        if (
+          value.highest_translated_timestamp > _highest_translated_timestamp) {
+            update_highest_translated_timestamp(
+              value.highest_translated_timestamp);
+        }
```
maybe we can make it a simple `_highest_translated_timestamp = std::max(value.highest_translated_timestamp, _highest_translated_timestamp)` ?

oleiman replied:
done

mmaslankaprv commented on src/v/datalake/translation/partition_translator_v2.cc:
```diff
@@ -0,0 +1,357 @@
+}
+
+ss::future<std::chrono::milliseconds>
+partition_translator::current_lag_ms() const {
```
i think this should take the highest offset stored in the log into account, otherwise for idle partitions the lag will always be growing with time

oleiman replied:
yeah, i agree. ~~not sure what the the function should be though 😕~~  never mind. misread "highest offset stored in the log" as "highest offset translated". yes, we just look at the data source and report 0ms if we're up to date.

oleiman replied:
done

mmaslankaprv commented on src/v/datalake/translation/partition_translator_v2.cc:
```diff
@@ -0,0 +1,357 @@
+        auto begin_offset = co_await _data_source->wait_for_data_to_translate(
+          last_translated_offset, _as);
+
+        // Notify the scheduler that there is some data to translate
+        _scheduler->notify_ready(id);
+
+        auto translate_f = co_await ss::coroutine::as_future<>(
+          translate_when_notified(begin_offset));
+
+        _inflight_translation_state.reset();
+
+        // Let the scheduler know we are done
+        _scheduler->notify_done(id);
```
nit: this section would benefit from comment, this seems to be crucial for the interaction between translator and scheduler

oleiman commented:
> The real test will be actually plugging it in I guess :)

Oh, it's plugged in. The test runs here are all exercising the new code. I just haven't removed the old version from the tree yet 😅

Feediver1 commented on src/v/config/configuration.cc:
```diff
@@ -3916,6 +3916,41 @@ configuration::configuration()
+      "Time given, in milliseconds, to a datalake translation as scheduled by "
+      "the datalake scheduler. Once a translation is scheduled, it will run "
+      "until either a) the time slice has elapsed or b) all pending records "
+      "on its souirce partition have been translated.",
```
```suggestion
      "Time given, in milliseconds, for a datalake translation as scheduled by "
      "the datalake scheduler. After a translation is scheduled, it will run "
      "until either a) the time specified has elapsed or b) all pending records "
      "on its source partition have been translated.",
```

oleiman replied:
done

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -237,25 +245,27 @@ ss::future<> datalake_manager::start() {
         }
     });
+
+    if (!_features->local().is_active(features::feature::datalake_iceberg_ga)) {
```
Curious why we need this? Translation scheduling seems like a local enough job that it's kinda surprising to see us waiting on other nodes to upgrade

Or is it that we require the state changes from the coordinator?

oleiman replied:
i hadn't given it much thought honestly, but I assumed there was something compat breaking in here...will have to give it a closer look

oleiman replied:
wasn't able to convince myself one way or another
@bharathv maybe recalls

bharathv replied:
This was added for custom_partitioning (CP) code..

```
    void reconcile_properties() override {
        auto old_invalid_action = _invalid_record_action;
        auto old_cp_enabled = _cp_enabled;
        auto old_target_lag = _target_lag;

        _invalid_record_action = compute_invalid_record_action();
        _cp_enabled = translation_task::custom_partitioning_enabled{.  <===
          _features.is_active(features::feature::datalake_iceberg_ga)};
```

CP is not enabled until all nodes upgrade to a version that support CP logic.  This provides an async notification to reset translators once that event happens.. 

Without this new code, the logic relied on short lived translation task logic [polling](https://github.com/redpanda-data/redpanda/blob/195956cf3bd7e89b5a6300700273195f5b90948c/src/v/datalake/translation/partition_translator.cc#L330) this variable at the begin of every translation loop.

oleiman replied:
cool, thx

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -264,128 +274,93 @@ ss::future<> datalake_manager::start() {
+                                 == model::iceberg_mode::disabled;
+    auto requires_active_translator = partition && topic_cfg
+                                      && !iceberg_disabled && is_leader;
+    if (translator_exists && !requires_active_translator) {
```
nit: maybe just me, but i think these conditions would be easier followed as

```
if !requires_active_translator:
  if translator_exists:
    remove translator
  return

if translator_exists:
  reconcile properties
  return

...set up translator
```

oleiman replied:
yeah i tend to agree. i've taken a light touch on the existing code so far, but happy to do this type of cleanup.

oleiman replied:
done

andrwng commented:
> Oh, it's plugged in. The test runs here are all exercising the new code. I just haven't removed the old version from the tree yet 😅

Oh hah, I skipped over the datalake_manager changes and never got back to them, and then thought they just didn't exist :)

oleiman committed:
datalake/t: partition_translator v2

Similar in function to the original partition_translator, but refactored
significantly to both a) work with the new scheduler and b) break some
coupling with various RPC interfaces to improve testability.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
config: Fix docstring on iceberg_target_lag_ms

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

bharathv reviewed (commented):
tysm for breaking it down into smaller commits 🙏

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/ea103594f1ffcdd4ea796bebf3c6f928e6229fbc..8e6b12e0a0840b224b06f20c507ea66fa2eb0b2f) many things
- removed old partition_translator from tree
- reworked lag timestamp to record last time we reached max translatable offset
- fixed some errors in config docs & removed path prefix config
- some strategically placed docstrings
- other minor cleanups, missed a small handful of others, forthcoming

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/8e6b12e0a0840b224b06f20c507ea66fa2eb0b2f..fbec37259b2ef39733abb2f01d50f8f052de5bff) rebase dev for merge conflict. See https://github.com/redpanda-data/redpanda/pull/25152#issuecomment-2690028244 for latest real diff

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/fbec37259b2ef39733abb2f01d50f8f052de5bff..a844b698ed8981e40878814ce0cc109c85c19157) few more cleanups

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/a844b698ed8981e40878814ce0cc109c85c19157..97476883d901ab1fd7b5a7f1547035dcc740f0a7) cruft

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/97476883d901ab1fd7b5a7f1547035dcc740f0a7..6d5dc0d39ab0104187ba1c48682659034dbf5051) rebase dev to fix merge conflict

oleiman commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62474-->
<!--26a4cc7f-5217-4ebd-a627-14d21be031b6-->
<!--20012d2f-95cb-44fc-be29-d71fc936865b-->
<!--title-->
## Retry command for [Build#62474](https://buildkite.com/redpanda/redpanda/builds/62474)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/mount_unmount_test.py::MountUnmountIcebergTest.test_simple_remount@{"cloud_storage_type":1}
tests/rptest/tests/datalake/mount_unmount_test.py::MountUnmountIcebergTest.test_simple_unmount@{"cloud_storage_type":1}
```
<!--/body-->

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -183,6 +183,14 @@ ss::future<> partition_translator::translate_until_stopped() {
+        // is NOT synchronized with outstanding commit operations. Therefore
+        // if we reach this point before the most recent batch of files has
+        // been committed, the commit lag metric will be out of sync at least
+        // until `wait_for_data_to_translate` wakes up.
```
think one improvement we could make is to make the wait time in wait_for_data_to_translate() bounded and redo the translation loop.

```
std::optional<kafka::offset>  begin_offset = co_await _data_source->wait_for_data_to_translatate(last_translated_offset, max_wait_time_s);

if (! begin_offset) {
  continue;
}
```

This keeps the metric stale for a bounded period of time.

oleiman replied:
ya great idea. 

>last_catchup_timestamp approximation,

we'll also update this each time we reenter the loop, so we'll get deterministic error bounds there too.

to the thing itself (sorry about the name), discussion is [here](https://github.com/redpanda-data/redpanda/pull/25152#discussion_r1973199231). Andrew's sugestion, but I'll try to concisely document the case for it sometime today.

bharathv reviewed (commented):
I didn't fully wrap my head around `last_catchup_timestamp` approximation, will take a (fresh) look at the last 4 commits tomorrow morning.

bharathv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -219,17 +218,6 @@ ss::future<iobuf> translation_stm::take_snapshot(model::offset) {
-      _highest_translated_offset,
-      new_offset);
-    _highest_translated_offset = new_offset;
-    _waiters_for_translated.notify(new_offset);
```
is the new code missing this notify() ?

oleiman replied:
ah, oops. overlooked that there was additional stuff happening in here.

oleiman replied:
fixed

oleiman commented:
CI Failures:
- [build-debug](https://buildkite.com/redpanda/redpanda/builds/62474#01954e6a-dd34-4af0-bce0-ed32dd9c87d5) there's a test that isn't in bazel so I missed in during development. it times out now but seems to pass in release mode, so probably just a question of cranking a deadline.

oleiman committed:
datalake/t: Timeout for partition_data_source::wait_for_data_to_translate

Previously we would poll indefinitely for translatable offsets. Instead, poll
for 30s and return nullopt if necessary.

This is helpful on idle or slow moving partitions, where quantities that depend
on periodic coordinator reconciliation could become stale.

commit_lag is one such quantity, and datalake_e2e_test is updated to account for
improved behavior.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

bharathv commented on src/v/datalake/translation/types.h:
```diff
@@ -11,20 +11,24 @@
     // to a cloud bucket and a subsequent notification to the datalake
     // coordinator of it's availablity.
     kafka::offset highest_translated_offset;
+    std::optional<model::timestamp> last_catchup_timestamp;
```
nit: a comment could be useful.

bharathv commented on src/v/datalake/translation/deps.h:
```diff
@@ -128,6 +128,24 @@ class data_source {
+     * returns optional<chrono::milliseconds>
+     *   - nullopt if iceberg disabled or stm is not in sync
+     *   - 0ms if last_translated_offset == max_offset_for_translation
+     *   - nullopt if we never caught up to max (TODO: not ideal)
```
I'm starting to wonder if we should use bytes_remaining as proxy metric for back pressure given this situation  🤔 .. Imagine translating a large partition with historic data and never catching up with the tip of the log and returning nullopt here and the backpressure controller doesn't know what to do, wonder if this case has already been discussed.

oleiman replied:
i think the hand wavy argument was that "never caught up to max" suggests that the partition has high throughput relative to avail resources and should be prioritized, but yea the historic data case is sort of pathological in that sense. besides which that state shouldn't overlap w/ the iceberg disables / stm sync case IMO.

> bytes_remaining

interesting. need to give it a bit more thought. is that readily available and/or cheap to compute?

mmaslankaprv replied:
I am thinking more and more about this, is it possible that we will never translate up to the `max_translatable_offset` but still we will be within the boundary of target lag ?

oleiman replied:
Yeah almost certainly

bharathv commented on src/v/datalake/translation/deps.cc:
```diff
@@ -181,9 +181,22 @@ class partition_data_source : public data_source {
+        // naively wait a few polling periods after some data appears, hopefully
+        // grabbing a few more batches before breaking back out to the
+        // translator
+        // TODO: probably choose a better value. configurable? or can we do it
```
I was thinking to do it bytes wise (new bytes to be translated since that affects batching) but this is ok too I guess since iceberg/parquet cares about records.

oleiman replied:
the code I wrote here was extremely broken so I dropped it from the PR in the interest of unblocking the next subdivision of remaining work. will revisit in the next PR, which will address other sort of batching oriented stuff.

oleiman commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62522-->
<!--2231df94-cf6e-47f8-8e61-718c5c12a817-->
<!--title-->
## Retry command for [Build#62522](https://buildkite.com/redpanda/redpanda/builds/62522)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/partition_movement_test.py::PartitionMovementTest.test_cross_core_movements@{"cloud_storage_type":1}
```
<!--/body-->

oleiman commented:
CI Failures:
- [debug-build](https://buildkite.com/redpanda/redpanda/builds/62522#019558d7-6f46-4cf2-8cbb-5dde7450cb13) nondeterministic but real failure in translator_test. looks like something is wonky on the `record_multiplexer` destruction path, leading to an unclosed ostream in parquet_writer. investigating [report](https://buildkite.com/organizations/redpanda/pipelines/redpanda/builds/62522/jobs/019558d7-6f46-4cf2-8cbb-5dde7450cb13/artifacts/01955960-f4c7-41b2-8b89-8f37c7907b87)
- [debug-ducktape](https://buildkite.com/redpanda/redpanda/builds/62522#0195591b-9412-4d90-8e7e-7af91981b986) `datalake/partition_movement_test::PartitionMovementTest.test_cross_core_movements` real, investigating

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/5da36f4181d86fa14f645db6a238ba52d1243c82..79f6b59856748c7a42a5c205c0903969cc30a0f0) fix a bug in topic revision accounting and add a code comment for `last_catchup_timestamp`.

From https://github.com/redpanda-data/redpanda/pull/25152#issuecomment-2693035265, the ducktape test should be fixed, but the fixture test is still broken. failure is nondeterministic (at least locally) triggering an assert in a seastar ostream implementation during parquet writer teardown.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/79f6b59856748c7a42a5c205c0903969cc30a0f0..9e582c2a0ba7fe826f2ff8abcd2dc38dcac1c2df) changes:
- fix eager as.check in `partition_translation_context::finish`
- adjust catchup timestamp calculation to check a cached "target offset"

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/9e582c2a0ba7fe826f2ff8abcd2dc38dcac1c2df..e64587015155043d86ab9b7be80e803f607b7f95) rebase dev to fix merge conflict

andrwng commented on src/v/datalake/translation/scheduling.h:
```diff
@@ -153,6 +153,18 @@ class translator {
+     * Heuristic for current lag of the translation, in milliseconds. More
+     * precisely, the wall-clock time elapsed since the _timestamp_ of the last
+     * translated record, as seen by translation_stm.
+     * (i.e. model::timestamp_clock::now() -last_translated_timestamp).
```
nit: needs an update?

oleiman replied:
yup good spot

andrwng commented on src/v/datalake/coordinator/translated_offset_range.h:
```diff
@@ -12,6 +12,7 @@
 #include "container/fragmented_vector.h"
 #include "datalake/coordinator/data_file.h"
 #include "model/fundamental.h"
```
nit: coordinator changes no longer needed?

andrwng commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -55,22 +56,30 @@ class translation_stm final : public raft::persisted_stm<> {
     void update_highest_translated_offset(kafka::offset new_offset);
 
     kafka::offset _highest_translated_offset{};
+    model::timestamp _last_catchup_timestamp{model::timestamp::min()};
```
Can you help clarify with a comment or a type change the difference between this being model::timestamp::min and having std::nullopt in translation_state? Feels like this should maybe be std::optional too?

oleiman replied:
hmm, yeah thought i changed this.

andrwng commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -130,7 +147,11 @@ ss::future<std::error_code> translation_stm::reset_highest_translated_offset(
-      make_translation_state_batch(new_translated_offset),
+      make_translation_state_batch(
+        new_translated_offset,
+        is_caught_up
+          ? std::make_optional<model::timestamp>(model::timestamp::now())
+          : std::nullopt),
```
Hmm I'm wondering if it makes sense to instead pass an optional::timestamp to reset_highest_translated_offset and rely on callers to give us the timestamp. Specifically, I'm wondering if we should cache the timestamp at which the "caught up" offset was cached, and then pass that timestamp here.

Like in the case of where translation interval is 5 minutes:
1. it is time T1, the max translatable offset is 100
2. we translate for 30 seconds
3. as we translate, data comes in at T2 and now the max translatable offset is 200
4. we finish translating to 100 and think we are caught up because max translated == 100, the time is now T3
5. we inform everyone we're now caught up as of T3, but really we should go ahead and consider offset 200's translation deadline to be T2 + 5m instead of T3 + 5m.
 
T1 + 5m would be inaccurate but perhaps is a better value to return than T3 + 5m, since it encourage the scheduler to schedule again sooner than T2 + 5m?

oleiman replied:
> T1 + 5m would be inaccurate but perhaps is a better value to return than T3 + 5m

yeah that's a fair point. i was trying to balance this with the totally idle topic case, where we want to keep that timestamp fresh on each iteration, but I think what you're suggesting covers that as long as we also update the timestamp on the timeout path through `wait_for_data`.

oleiman replied:
actually I think the idle topic case can be covered when we compute lag. so really we do just cache the last time we updated the target offset and replicate that if & when the highest translated offset exceeds it.

oleiman replied:
I wonder whether it makes sense to update the target _only_ if we've met the previous one. if we do that, we'll producer a 'worse' underestimate of the T2 records' deadline, but this way we never get into a situation where we're always chasing the tip of the log (and never recording a timestamp). wdyt?

andrwng commented on src/v/datalake/translation/deps.h:
```diff
@@ -128,6 +128,24 @@ class data_source {
+     * returns optional<chrono::milliseconds>
+     *   - nullopt if iceberg disabled or stm is not in sync
+     *   - 0ms if last_translated_offset == max_offset_for_translation
+     *   - nullopt if we never caught up to max (TODO: not ideal)
```
Is nullopt supposed to be in this bullet list twice? As a caller that seems confusing. Do we need another state to indicate errors? Maybe a result type?

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -284,10 +287,21 @@ class partition_data_source : public data_source {
     }
 
+    void update_translation_target() {
+        _maybe_translation_target = max_offset_for_translation();
```
Left a comment elsewhere, and it's probably not a huge deal, but I'm wondering if it makes sense to also cache the timestamp when this was called with a new offset, and then pass that into `reset_highest_translated_offset()` rather than calling `now()` at replication time. I guess the win there is that if data comes in while we're translating, we don't miss out accounting for that new data.

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -184,13 +184,20 @@ ss::future<> partition_translator::translate_until_stopped() {
+            last_translated_offset, _as);
+
+        if (!maybe_begin_offset.has_value()) {
+            continue;
```
nit: I think a comment would be helpful here to mention the motivation behind continuing (like as is, it's totally unclear that this is important for metrics, but i also imagine this could be useful if there's some mismatch between the coordinator and the translator?)

oleiman replied:
Also an opportunity to keep the catchup timestamp up-to-date so lag doesn't grow unbounded on an idle topic, but i will add a comment.

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -336,8 +336,8 @@ datalake_manager::handle_translator_state_change(const model::ntp& ntp) {
         local_path{_writer_scratch_space},
-        partition->ntp(),
-        partition->get_revision_id(),
+        data_src->ntp(),
+        data_src->topic_revision(),
```
>should track topic_revision_id
>Rather than raft group revision.

This makes sense. Did this surface as a test failure?

oleiman replied:
yeah. the integration test mentioned here: https://github.com/redpanda-data/redpanda/pull/25152#issuecomment-2693035265

bharathv commented:
nit: squash the last 3 (or last 2 if you'd prefer) bug fixes into their corresponding commits?

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/346c0f1cc64c77a924873ce42e279b96ec33ad9e..4844d4c645e1bf9c78f3cf2cac7b022b392cd3be) CR comments. reframed "catchup timestamp" as an approximation of last translated timestamp, hopefully an underestimate in the common case.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/4844d4c645e1bf9c78f3cf2cac7b022b392cd3be..502834a7ee95102e1280a486b2c386a9656a08cd) rebase `dev` to resolve conflict

andrwng commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -187,6 +210,7 @@ ss::future<> translation_stm::apply_raft_snapshot(const iobuf&) {
     // with the snapshot.
     vlog(_log.debug, "Applying raft snapshot, resetting state");
     _highest_translated_offset = kafka::offset{};
+    _last_translated_timestamp = model::timestamp::min();
```
What does this mean vs std::nullopt? I'm still kind of confused what the difference is between min and nullopt are, and it feels like an easy footgun. Do we need to mix them? If so, can you leave a comment in the declaration about its semantics?

oleiman replied:
> what does this mean vs std::nullopt

means nothing / is a bug. missed a spot scrubbing the diff.

oleiman replied:
ah, this is in the raft snapshot applier. gonna address in a quick followup.

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,545 @@
+      model::timeout_clock::duration timeout,
+      ss::abort_source& as) final {
+        std::optional<model::timestamp> new_ts{};
+        if (_translation_target && new_offset >= _translation_target->offset) {
+            new_ts.emplace(_translation_target->ts);
+        }
```
Maybe this should only be updated on success?

oleiman replied:
not sure I follow. are you suggesting something like [this](https://github.com/redpanda-data/redpanda/pull/25152#discussion_r1978555852)?

andrwng replied:
Oh 🤦 when I wrote this i thought this was a member variable. Nevermind!

andrwng commented on src/v/datalake/translation/deps.h:
```diff
@@ -21,4 +34,183 @@ class noop_mem_tracker : public writer_mem_tracker {
+     * current_lag_ms - Approximation of current translation lag with respect to
+     * some data source.
+     *
+     * Calulated as the difference between current system time and a replicated
+     * underestimate for the timestamp of the last offset we translated. The
```
nit: this doesn't seem like it can be true, given the waiting interval expands our error bounds. Not sure it's an important enough gotcha to update the comment, but something to think about

oleiman replied:
yeah, i can't stop thinking about it 🙃. the thing I'm trying to express is that this timestamp is the value of a semi-arbitrary function `F(Offset) -> timestamp` applied to _some_ offset <= the highest translated. But I think that's even more cryptic as a code comment.

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,545 @@
+        return kafka::prev_offset(model::offset_cast(lso.value()));
+    }
+
+    ss::future<std::error_code> update_highest_translated_offset(
```
nit: might be nice to call it replicate_highest_translated_offset() or something so it's harder to confuse with the other update_highest_translated_offset()

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -10,12 +10,545 @@
+      model::timeout_clock::duration timeout,
+      ss::abort_source& as) final {
+        std::optional<model::timestamp> new_ts{};
+        if (_translation_target && new_offset >= _translation_target->offset) {
```
nit: I think this makes sense, but I think it'll be easy to forget the context about why this looks the way it does. I think a comment here would be useful

andrwng reviewed (approved):
None of my comments are blocking
